### ISSUE TITLE
NullableDereference ignores methods that throw if null in call path

### ISSUE THREAD

**Issue by bobtiernay-okta** [google/error-prone#1253]
### Description of the problem / feature request:

The following common [AssertJ](http://joel-costigliola.github.io/assertj/) `isNotNull()` assertion will not pass `NullableDereference`:

```java
Integer value = null;
assertThat(value)
   .isNotNull()
   .isEqualTo(1);
```

Similarly, neither will this:

```java
Integer x = null;
Integer y = Preconditions.checkNotNull(x);
int result = 1 + y;
```

And this:

```java
class Foo {
  Integer value;
  
  String get() {
      Preconditions.checkNotNull(value);
      return value.toString();
  }
}
```

### Feature requests: what underlying problem are you trying to solve with this feature?

In order to prevent false positives, it null analysis should examine the call path for null checks.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

See above.

### What version of Error Prone are you using?

2.3.3

###  Have you found anything relevant by searching the web?

No.

**Comment by kevin1e100** [comment]
Thanks for your report.  Note this checker is experimental and incomplete at this point; please don't rely on it.

There's actually two issues here:
1. Based on checkNotNull's annotations, the checker should warn when passing null to it.
1. Ideally as mentioned above the checker would assume checkNotNull's result to be non-null.

I say "ideally" b/c what's being asked here is ironically somewhat difficult to do once there is an error, which is the case here.

You should be able to side-step this issue by using `Verify.verifyNotNull` instead.  That method is more appropriate for "sanitizing" possibly-null values, as its annotations allow `null` values to be passed in.  

NullableDereference uses available annotations to do its work and the two methods are annotated [subtly](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Preconditions.java) [differently](https://github.com/google/guava/blob/master/guava/src/com/google/common/base/Verify.java) to allow nullable inputs only for `verifyNotNull`.

**Comment by cushon** [comment]
d73fa691c8a60cf88934b585b4d908c9f07c868b
